libscript/src/script-validate.hpp

```diff
@@ -0,0 +1,299 @@
+	void ReportInvalidAddress(void);
+	void ReportInvalidRegister(void);
+	void ReportInvalidValue(void);
+	void ReportInvalidDefinition(void);
+	void ReportInvalidHandler(void);
+	void ReportInvalidFetchable(void);
+	void ReportInvalidVariable(void);
+	
+private:
+	// Whilst validation of bytecode is successful, the error var remains
+	// false. When it becomes true, it means the bytecode at 'current_address'
+	// within the handler is invalid.
+	bool m_error;
+	
+	// The module containins the handler being validated.
+	MCScriptModuleRef m_module;
+	
+	// The handler being validated.
+	MCScriptHandlerDefinition *m_handler;
+	
+	// The offset from the beginning of the module's bytecode to the current
+	// instruction being validated.
+	uindex_t m_current_address;
+	
+	// The current instructions argument count.
+	uindex_t m_argument_count;
+	
+	// The current instructions list of (decoded) arguments.
+	uindex_t m_arguments[256];
+	
+	// The total number of registers required to run the bytecode.
+	uindex_t m_register_limit;
+};
+
+inline uindex_t MCScriptValidateContext::GetAddress(void) const
+{
+	return m_current_address;
+}
+
+inline uindex_t MCScriptValidateContext::GetArity(void) const
+{
+	return m_argument_count;
+}
+
+inline uindex_t MCScriptValidateContext::GetArgument(uindex_t p_index) const
+{
+	__MCScriptAssert__(p_index < m_argument_count, "invalid argument index");
+	return m_arguments[p_index];
+}
+
+inline index_t MCScriptValidateContext::GetSignedArgument(uindex_t p_index) const
+{
+	uindex_t t_unsigned_value = GetArgument(p_index);
+	
+	index_t t_value;
+	if ((t_unsigned_value & 1) == 0)
+		t_value = (signed)(t_unsigned_value >> 1);
+	else
+		t_value = -(signed)(t_unsigned_value >> 1);
+	
+	return t_value;
+}
+
+inline MCScriptDefinitionKind MCScriptValidateContext::GetEffectiveKindOfDefinition(uindex_t p_index) const
+{
+	__MCScriptAssert__(p_index < m_module->definition_count, "invalid definition index");
+	
+	if (m_module->definitions[p_index]->kind == kMCScriptDefinitionKindExternal)
+	{
+		MCScriptExternalDefinition *t_ext_def =
+			static_cast<MCScriptExternalDefinition *>(m_module->definitions[p_index]);
+		
+		return m_module->imported_definitions[t_ext_def->index].kind;
+	}
+	
+	return m_module->definitions[p_index]->kind;
+}
+
+inline void MCScriptValidateContext::CheckArity(uindex_t p_expected_arity)
+{
+	if (m_error)
+		return;
+	
+	if (GetArity() != p_expected_arity)
+	{
+		ReportInvalidArity();
+		return;
+	}
+}
+
+inline void MCScriptValidateContext::CheckAddress(uindex_t p_address)
+{
+	if (m_error)
+		return;
+	
+	if (p_address < m_handler->start_address)
+	{
+		ReportInvalidAddress();
+		return;
+	}
+	
+	if (p_address >= m_handler->finish_address)
+	{
+		ReportInvalidAddress();
+		return;
+	}
+}
+
+inline void MCScriptValidateContext::CheckRegister(uindex_t p_register)
+{
+	if (m_error)
+	{
+		return;
+	}
+	
+	if (p_register >= m_register_limit)
+	{
+		m_register_limit = p_register + 1;
+	}
```

Comment:
Bytecode blocks only allow variable definitions which are always allocated a register in order from the sequence of definitions. The actual number of registers needed is computed as the maximum of all register indicies in the bytecode of the handler... This means it would be possible to overflow that - the maximum register index needs to be UINDEX_MAX-1. I'm thinking of making it so that the compiler emits the number of registers needed for a handler, and then the validation check is that there are no indicies greater than that and *all* indicies are referenced at least once in the bytecode. This concern goes away completely when we move to a stack-based format though, so I'll just make CheckRegister check that p_register < UINDEX_MAX for now.